### PR TITLE
Verify sample code provided in issue 4060

### PR DESCRIPTION
**Pull request by Shane32**

### CODE REVIEW

File: src/GraphQL.Tests/Bugs/Issue4060.cs [hunk 1]
```diff
@@ -0,0 +1,228 @@
...
+                value = value2 ?? ImmutableDictionary<string, ArgumentValue>.Empty;
+
+                // Cache the current field and its arguments to avoid recomputation.
+                lock (this)
+                {
+                    _lastField = key;
+                    _lastArgs = value;
+                }
+
+                return true;
```
Comment (gao-artur) [inline src/GraphQL.Tests/Bugs/Issue4060.cs:220]: The method never returns `false`... Should the method signature change?
```c#
public bool TryGetValue(GraphQLField key, out IDictionary<string, ArgumentValue> value)
```
To
```c#
public IDictionary<string, ArgumentValue> GetValue(GraphQLField key)
```

File: src/GraphQL.Tests/Bugs/Issue4060.cs [hunk 2]
```diff
@@ -0,0 +1,228 @@
...
+            // Dictionary that maps GraphQLField nodes to their corresponding FieldType objects.
+            // This is used to store field definitions as the AST is traversed.
+            private readonly Dictionary<GraphQLField, FieldType> _fields = new();
+
+            // `EnterAsync` is called when entering a node in the AST. 
+            // If the node is a `GraphQLField`, the corresponding `FieldType` is fetched and stored in the `_fields` dictionary.
+            public ValueTask EnterAsync(ASTNode node, ValidationContext context)
+            {
+                // Check if the current AST node is a `GraphQLField`.
+                if (node is GraphQLField field)
```
[the remark this answers is not in the document]
Comment (Shane32) [reply src/GraphQL.Tests/Bugs/Issue4060.cs:97]: the comments were yeah, i can remove the comments if you like.

File: src/GraphQL.Tests/Bugs/Issue4060.cs [hunk 3]
```diff
@@ -0,0 +1,228 @@
...
+
+            // Constructor that initializes the `FieldArgumentDictionary` with an execution context and a field mapping.
+            public FieldArgumentDictionary(IExecutionContext executionContext, IDictionary<GraphQLField, FieldType> fieldMappings)
+            {
+                _executionContext = executionContext;
+                _fieldMappings = fieldMappings;
+            }
+
+            // Indexer to access argument values for a given `GraphQLField`.
+            public IDictionary<string, ArgumentValue> this[GraphQLField key]
```
[the remark this answers is not in the document]
Comment (Shane32) [reply src/GraphQL.Tests/Bugs/Issue4060.cs:159]: yes

File: src/GraphQL.Tests/Bugs/Issue4060.cs [hunk 4]
```diff
@@ -0,0 +1,228 @@
...
+
+            // Returns an enumerator, though it is not implemented here.
+            public IEnumerator<KeyValuePair<GraphQLField, IDictionary<string, ArgumentValue>>> GetEnumerator() => throw new NotImplementedException();
+
+            // `TryGetValue` is the core logic for retrieving argument values for a given field.
+            public bool TryGetValue(GraphQLField key, out IDictionary<string, ArgumentValue> value)
+            {
+#pragma warning disable RCS1059 // Avoid locking on publicly accessible instance
+                // Locking is used to ensure thread safety when caching field-argument mappings.
+                lock (this)
```
[the remark this answers is not in the document]
Comment (Shane32) [reply src/GraphQL.Tests/Bugs/Issue4060.cs:186]: well, yeah, although i don't really see this as an applicable warning.

File: src/GraphQL.Tests/Bugs/Issue4060.cs (hunk 1 above)
Comment (Shane32) [reply src/GraphQL.Tests/Bugs/Issue4060.cs:220]: no, this method signature comes from `IReadOnlyDictionary...`

**Comment by Shane32** [comment]
Note: this code is not intended to be added into GraphQL.NET proper - it's just testing some sample code.  So I did not clean it up as I would normally do.

File: src/GraphQL.Tests/Bugs/Issue4060.cs (hunk 2 above)
[the remark this answers is not in the document]
Comment (gao-artur) [reply src/GraphQL.Tests/Bugs/Issue4060.cs:97]: There are so many comments that it was easier just to skip them and read the code to understand the logic. I would remove the "obvious" comments to make the code more readable.

**Comment by gao-artur** [comment]
Noted.

File: src/GraphQL.Tests/Bugs/Issue4060.cs (hunk 4 above)
[the remark this answers is not in the document]
Comment (Shane32) [reply src/GraphQL.Tests/Bugs/Issue4060.cs:186]: This is a sample of code for a developer to utilize.  If they were locking on `ArgumentValues`, they should change the sample.  GraphQL.NET does not lock on `ArgumentValues`.

[the remark this answers is not in the document]
Comment (Shane32) [reply src/GraphQL.Tests/Bugs/Issue4060.cs:186]: (I will change the sample anyway though.)

File: src/GraphQL.Tests/Bugs/Issue4060.cs [hunk 5]
```diff
@@ -0,0 +1,228 @@
...
+            }
+
+            // Indexer to access argument values for a given `GraphQLField`.
+            public IDictionary<string, ArgumentValue> this[GraphQLField key]
+            {
+                get
+                {
+                    // If the field exists in the dictionary, return its argument values.
+                    if (TryGetValue(key, out var value))
+                        return value!;
```
[the remark this answers is not in the document]
Comment (Shane32) [reply src/GraphQL.Tests/Bugs/Issue4060.cs]: fixed

File: src/GraphQL.Tests/Bugs/Issue4060.cs (hunk 2 above)
[the remark this answers is not in the document]
Comment (Shane32) [reply src/GraphQL.Tests/Bugs/Issue4060.cs:71]: i asked chatgpt to remove all obvious comments, and it removed them all so 🤷‍♂️  they are all gone now

File: src/GraphQL.Tests/Bugs/Issue4060.cs (hunk 4 above)
[the remark this answers is not in the document]
Comment (Shane32) [reply src/GraphQL.Tests/Bugs/Issue4060.cs]: changed

**Review by gao-artur — APPROVED** [review]

**Pull request merged**